web/src/lib/utils/cast/gcast-destination.svelte.ts

```diff
@@ -0,0 +1,264 @@
+    // in the form of a global window event. We will give Chrome 3 seconds to respond
+    // or we will mark the destination as unavailable
+
+    const callbackPromise: Promise<boolean> = new Promise((resolve) => {
+      if (this.isAvailable) {
+        console.debug('GCast API already available');
+        resolve(true);
+        return;
+      }
+
+      window['__onGCastApiAvailable'] = (isAvailable: boolean) => {
+        resolve(isAvailable);
+      };
+
+      if (!document.querySelector(`script[src="${FRAMEWORK_LINK}"]`)) {
+        const script = document.createElement('script');
+        script.src = FRAMEWORK_LINK;
+        document.body.append(script);
+      }
+    });
+
+    const timeoutPromise: Promise<boolean> = new Promise((resolve) => {
+      setTimeout(
+        () => {
+          resolve(false);
+        },
+        Duration.fromObject({ seconds: 3 }).toMillis(),
+      );
+    });
+
+    this.isAvailable = await Promise.race([callbackPromise, timeoutPromise]);
+
+    console.debug('GCast API available:', this.isAvailable);
+
+    if (this.isAvailable) {
+      const castContext = cast.framework.CastContext.getInstance();
+      this.remotePlayer = new cast.framework.RemotePlayer();
+
+      castContext.setOptions({
+        receiverApplicationId: chrome.cast.media.DEFAULT_MEDIA_RECEIVER_APP_ID,
+        autoJoinPolicy: chrome.cast.AutoJoinPolicy.ORIGIN_SCOPED,
+      });
+
+      //
+      // setup the event listeners
+      //
+
+      castContext.addEventListener(
+        cast.framework.CastContextEventType.SESSION_STATE_CHANGED,
+        this.onSessionStateChanged.bind(this),
+      );
+
+      castContext.addEventListener(
+        cast.framework.CastContextEventType.CAST_STATE_CHANGED,
+        this.onCastStateChanged.bind(this),
+      );
+
+      const remotePlayerController = new cast.framework.RemotePlayerController(this.remotePlayer);
+      remotePlayerController.addEventListener(
+        cast.framework.RemotePlayerEventType.ANY_CHANGE,
+        this.onRemotePlayerChange.bind(this),
+      );
+    }
+
+    return this.isAvailable;
+  }
+
+  async loadMedia(mediaUrl: string, sessionKey: string, reload: boolean = false): Promise<void> {
+    if (!this.isAvailable || !this.isConnected) {
+      console.debug('Cannot load media, cast destination is not available or not connected');
+      return;
+    }
+
+    if (!this.session) {
+      console.debug('No active session, cannot load media');
+      return;
+    }
+
+    if (this.currentUrl === mediaUrl && !reload) {
+      console.debug('Already casting this media, skipping load');
+      return;
+    }
+
+    // we need to send content type in the request
+    let contentType: string | null = null;
+
+    // in the future we can swap this out for an API call to get image metadata
+    await fetch(mediaUrl, { method: 'HEAD' }).then((response) => {
+      contentType = response.headers.get('content-type');
+    });
```

Comment:
Maybe I am too tired right now, but what are you awaiting here?

Reply:
This is to grab the content-type for the thumbnail. @jrasm91 would it be too hard to get the thumbnail contentType with the asset info, would come in handy on mobile to keep it using openapi calls.